File: leetcode/SpiralMatrixII/generateMatrix.java
public class Solution {
    public int[][] generateMatrix(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int layer = 0, num = 0, matrix[][] = new int[n][n];
        for (; layer < n / 2; layer++) {
            for (int i = layer; i < n - layer - 1; i++) matrix[layer][i] = ++num;
            for (int i = layer; i < n - layer - 1; i++) matrix[i][n - layer - 1] = ++num;
            for (int i = n - layer - 1; i > layer; i--) matrix[n - layer - 1][i] = ++num;
            for (int i = n - layer - 1; i > layer; i--) matrix[i][layer] = ++num;
        }
        if (n % 2 == 1) {
            matrix[layer][layer] = ++num;
        }
        return matrix;
    }
}

File: leetcode/SpiralMatrixII/generateMatrix.cpp
Spiral Matrix II
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int v = 1;
        vector<vector<int>> ans(n, vector<int>(n));
        for (int i = 0; i <= (n - 1) / 2; i++) {
            int len = n - 2 * i;
            for (int j = 0; j < len - 1; j++, v++) {
                ans[i][i + j] = v;
                ans[i + j][n - 1 - i] = (len - 1) + v;
                ans[n - 1 - i][n - 1 - i - j] = (len - 1) * 2 + v;
                ans[n - 1 - i - j][i] = (len - 1) * 3 + v;
            }
            if (len == 1) ans[i][i] = v;
            v += (len - 1) * 3;
        }
        return ans;
    }
};
